Api gui howto guide

brynnz22 opened the pull request:
@mslarae13 @jkelliher-github here is a first draft of some api gui documentation aimed at data consumers. It just includes information and examples for the find and metadata endpoints. Please feel free to give lots of feedback or if its easier to make any edits directly! 

Thanks!

brynnz22 commented:
@dwinston @aclum  Please let me know if there is any inaccuracies or feedback you have! Thanks!

dwinston commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,165 @@
+
+## Retrieving and Submitting Metadata using the ___Find___ and ___Metadata___ API Endpoints
+
+Metadata describing NMDC data (e.g. studies, biosamples, data objects, etc.) may be retrieved or submittied with GET and POST requests, respectively, using the **[NMDC API Graphical User Interface (GUI)](https://api.microbiomedata.org/docs#/)** The API GUI provides an user interface for programmatic access to the NMDC data portal without needing to use the Command Line.
```
needs a basic editing pass, e.g. `submittied` -> `submitted`, but I won't point out any more such trivial changes in this review.

eecavanna replied:
Fixed in a05a20ed8ee3a92fe14affe9998f6d5839528933

brynnz22 replied:
Thanks for fixing these @eecavanna!

dwinston commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,165 @@
+
+## Retrieving and Submitting Metadata using the ___Find___ and ___Metadata___ API Endpoints
+
+Metadata describing NMDC data (e.g. studies, biosamples, data objects, etc.) may be retrieved or submittied with GET and POST requests, respectively, using the **[NMDC API Graphical User Interface (GUI)](https://api.microbiomedata.org/docs#/)** The API GUI provides an user interface for programmatic access to the NMDC data portal without needing to use the Command Line.
```
> user interface for programmatic access

"programmatic access" is generally non-graphical and non-interactive, i.e. via python scripting, shell scripting with `curl`, etc. I'd emphasize that what this API GUI provides is a guided interface for (1) performing highly granular and targeted queries directly -- ones that may not be directly supported by the Data Portal UI (yet), and (2) exploring query capabilities interactively and in the process providing code snippets to adapt a data-consumer workflow to using programmatic access, i.e. a user can copy the formed URLs, `curl` commands, etc. to help them construct scripts.

brynnz22 replied:
Thank you both. I included your feedback in [7372ee3f3953633f8540a8860487803c316e9a99](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/7372ee3f3953633f8540a8860487803c316e9a99)

dwinston commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,165 @@
+
+Metadata describing NMDC data (e.g. studies, biosamples, data objects, etc.) may be retrieved or submittied with GET and POST requests, respectively, using the **[NMDC API Graphical User Interface (GUI)](https://api.microbiomedata.org/docs#/)** The API GUI provides an user interface for programmatic access to the NMDC data portal without needing to use the Command Line.
+
+Requests can include various parameters to filter, sort, and organize the requested information. Attribute names in the parameters will vary depending on the collection. The required syntax of the parameters will also vary, depending on if it is a ___find___ or a ___metadata___ endpoint. ___Find___ endpoints are designed to use more compact syntax (for example, filtering biosamples for an "Ecosys ...
```
optional: you can point to https://docs.openalex.org/how-to-use-the-api/get-lists-of-entities/filter-entity-lists as prior art for the compact syntax.

brynnz22 replied:
Added it! [9eb3bc4](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/9eb3bc4afcb79d30fdd78738a054700b0beea4eb) Thank you!

eecavanna reviewed (commented):
Looks great so far! I've found it easy to read and digest (as someone who does have experience using Swagger UI, but doesn't have much—in some cases, any—experience with the endpoints covered).

I have only reviewed up to the beginning of the "Metadata Endpoints" section. I'll continue from there ~~another day~~ later today.

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+| filter | Allows conditions to be set as part of the query, returning only results that satisfy the conditions | [MongoDB-like query language](https://www.mongodb.com/docs/manual/tutorial/query-documents/). All strings should be in double quotation marks. | `{"lat_lon.latitude": {"$gt": 45.0}, "ecosystem_category": "Plants"}` | 
+| max_page_size | Specifies the maximum number of documents returned at a time | Integer | `25`
+| page_token | Specifies the token of the page to return. If unspecified, the first page is returned. To retrieve a subsequent page, the value received as the `next_page_token` from the bottom of the previous results can be provided as a `page_token`. ![next_page_token](../_static/images/howto_guides/api_gui/metadata_page_token_param.png) | String | `nmdc:sys0ae1sh583`
+| projection | Indicates the desired fields to be included in the response. Helpful for trimming down the returned results | Comma separated string of field names that correspond to a `collection_name`. | `name, ecosystem_type` |
```
The wording of "field names that correspond to a `collection_name`" was a little confusing to me. Here's an alternative I find a little more clear, but not as concise.

```diff
- Comma separated string of field names that correspond to a `collection_name`
+ Comma-separated list of fields that belong to the documents in the collection being queried
```

brynnz22 replied:
Thank you! I added your suggestion. I also change "fields" to "attributes". I was trying to stay consistent with what we call these and I think I landed on attributes, but there might be some "fields" sprinkled in there. See commit [f0f351a](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/f0f351ab5090198cb21d375dfc9fb6e503f6376c)

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+<br/>
+
+![metadata post changesheets validate](../_static/images/howto_guides/api_gui/metadata_post_changesheets_validate.png)
+A csv or tsv file can be validated against the NMDC schema using the `POST /metadata/changesheets:validate` endpoint in order to update an already existing record in the portal. Please see an [example changesheet](https://github.com/microbiomedata/nmdc-runtime/blob/main/metadata-translation/notebooks/data/changesheet-without-separator3.tsv). The file should include four columns:
```
I recommend capitalizing "CSV" and "TSV"

```diff
- csv or tsv
+ CSV or TSV
```

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+| doc_id | The unique identifier of the item being requested. For example, the identifier of a biosample or an extraction | Curie e.g. `prefix:identifier` | `nmdc:bsm-11-ha3vfb58` |<br/>
+<br/>
+
+The __metadata__ endpoints allow users to retrieve metadata from the data portal using the various `GET` endpoints that are slightly different than the __find__ endpoints, but some can be used similarily. They also include the ability to `POST` metadata changes to the data portal by allowing the validation and submission of change sheets or JSON files. Change sheets are spreadsheets that specify  ...
```
"Change sheets" is spelled with a space in this paragraph, but without a space in the paragraph below.

brynnz22 replied:
@mslarae13 and I discussed taking the change sheets endpoints out of this documentation since this is for users and adding to Donny's docs that is more for developers.

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+3. `attribute`: the attribute (or field/slot) that will be updated (e.g. `name` or `ecosystem_category`, etc.)
+4. `value`: the new value that will be inserted or that will replace the old value.
+
+Please note that if changes are made to multivalued attributes that have a "structured" value, e.g. [air_temp_regm](https://microbiomedata.github.io/nmdc-schema/air_temp_regm/) has a range of [TextValue](https://microbiomedata.github.io/nmdc-schema/TextValue/), set the `value` to a variable, then set the `id` of the next line to the variable with the `value` set to what will get populated. See [e ...
```
I recommend adding a trailing period.

```diff
- please see [Authoring Changesheets](https://microbiomedata.github.io/nmdc-runtime/howto-guides/author-changesheets/)<br/>
+ please see [Authoring Changesheets](https://microbiomedata.github.io/nmdc-runtime/howto-guides/author-changesheets/).<br/>
```

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+<br/>
+
+![metadata get collection name](../_static/images/howto_guides/api_gui/metadata_get_collection_name.png)
+The `GET /nmdcschema/{collection_name}` endpoint is a general purpose way to retrieve metadata about a specified collection given user-provided filter and projection criteria. Please see the [Collection Names](https://microbiomedata.github.io/nmdc-schema/Database/) that may be retrieved. Please note that only one collection set may be retrieved at a time.<br/>
```
When I read this sentence, I wondered what a collection set was. Based on my interpretation of the rest of the paragraph, here's a version of the sentence that doesn't use that term:

```diff
- Please note that only one collection set may be retrieved at a time.
+ Please note that metadata may only be retrieved about one collection at a time.
```

brynnz22 replied:
Thanks! I replaced this sentence with your suggestion: [6af0528](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/6af0528dcecd180368a851e7e571c45fbc41ece7)

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+If both the identifier and the collection name of the desired record is known, the `GET /nmdcschema/{collection_name}/{doc_id}` can be used to retrieve the record. The projection parameter is optionally available for this endpoint to retrieve only desired attributes from a record. Please note that only one record can be retrieved at one time using this method.<br/>
+<br/>
+
+#### Metadata Endpoints Example 1: Get all of the biosamples part of the 1000 Soils Research Campaign Study sampled from Colorado
```
I recommend adding "that are" between "biosamples" and "part"

```diff
- Get all of the biosamples part of
+ Get all of the biosamples that are part of
```

brynnz22 replied:
Added! [f7b1c1d](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/f7b1c1d4da6321c512d6eb0aff52f06949c7bd55)

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+4. Enter in the parameters in the **`GET /nmdcschema/{collection_name}`** endpoint. For this example, we will input `biosample_set` into the **collection_name** parameter and `{"part_of": "nmdc:sty-11-28tm5d36", "geo_loc_name.has_raw_value": {"$regex": "Colorado"}}` into the **filter** parameter. See the [Biosample Class](https://microbiomedata.github.io/nmdc-schema/Biosample/) in the NMDC Schema ...
+5. Click **Execute**
+![metadata example step4](../_static/images/howto_guides/api_gui/metadata_example_step4and5.png)
+6. View the results in a json format, available to download by clicking **Download**, or copy the results by clicking the clipboard icon in the bottom right corner of the response. In this case two studies were retrieved. Note that the curl and request URL are provided as well.
```
I recommend capitalizing "json"

```diff
- json
+ JSON
```

eecavanna reviewed (commented):
I finished reviewing the entire document as of commit 454fd9918deb6f43d660e226d4ca14b1abd756ff (and as of commit a05a20ed8ee3a92fe14affe9998f6d5839528933).

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+<br/>
+
+![find get activities](../_static/images/howto_guides/api_gui/find_get_activities.png)
+The `GET /activities` endpoint is a general way to fetch metadata about various activities (e.g. metagenome assembly, natural organic matter analysis, library preparation, etc.). Any "slot" (a.k.a. attribute) for [WorkflowExecutionActivty](https://microbiomedata.github.io/nmdc-schema/WorkflowExecutionActivity/) or [PlannedProcess](https://microbiomedata.github.io/nmdc-schema/PlannedProcess/)  may ...
```
Regarding the final sentence in this paragraph:

> Any "slot"...

I had a hard time understanding that particular sentence. I wonder whether it could be split into two.

brynnz22 replied:
I fixed this. I think there was a similar comment above. Let me know if it is still difficult to understand.

mslarae13 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+
+## Retrieving and Submitting Metadata using the ___Find___ and ___Metadata___ API Endpoints
+
+Metadata describing NMDC data (e.g. studies, biosamples, data objects, etc.) may be retrieved or submitted with GET and POST requests, respectively, using the **[NMDC API Graphical User Interface (GUI)](https://api.microbiomedata.org/docs#/)**. The API GUI provides a user interface for programmatic access to the NMDC data portal without needing to use the Command Line.
```
retrieved or submitted... do we need submitted for a user documentation?

brynnz22 replied:
Removed all the metadata "submission" and post topics. See commit [c96e9ee](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/c96e9ee6fa5dc0e3579412152b1e103909a2a897)

mslarae13 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+
+#### ___Metadata___ Endpoints
+
+The [metadata endpoints](https://api.microbiomedata.org/docs#/metadata) can be used to get and filter metadata from collection set types (including studies, biosamples, activites, and data objects as discussed in the __find__ section), as well as validate and submit updates to existing metadata to the data portal as a spreadsheet or json file. 
```
same question .... and maybe I'm forgetting what our target was. If it's just users ... we don't need the "as well as validate and submit updates to existing metadata to the data portal as a spreadsheet or json file" part.

brynnz22 replied:
Removed! [c96e9ee](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/c96e9ee6fa5dc0e3579412152b1e103909a2a897)

mslarae13 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+
+The [metadata endpoints](https://api.microbiomedata.org/docs#/metadata) can be used to get and filter metadata from collection set types (including studies, biosamples, activites, and data objects as discussed in the __find__ section), as well as validate and submit updates to existing metadata to the data portal as a spreadsheet or json file. 
+
+The syntax for the filter parameter of the __metadata__ endpoints is slightly different than that of the __find__ endpoints using [MongoDB-like language querying](https://www.mongodb.com/docs/manual/tutorial/query-documents/) instead of the compact syntax the __find__ endpoints use. The applicable parameters of the __metadata__ endpoints, with acceptable syntax and examples, are in the table belo ...
```
this is a little hard to read... recommend 

Unlike the compact syntax used in the __find__ endpoints, the syntax for the filter parameter of the __metadata__ endpoints use [MongoDB-like language querying](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The applicable parameters of the __metadata__ endpoints, with acceptable syntax and examples, are in the table below.

brynnz22 replied:
Thanks! I replaced this paragraph with your suggestion [8ec86c6](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/8ec86c6cdde32a6d109ee0300067b2d2ebc3e89b)

mslarae13 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+| doc_id | The unique identifier of the item being requested. For example, the identifier of a biosample or an extraction | Curie e.g. `prefix:identifier` | `nmdc:bsm-11-ha3vfb58` |<br/>
+<br/>
+
+The __metadata__ endpoints allow users to retrieve metadata from the data portal using the various `GET` endpoints that are slightly different than the __find__ endpoints, but some can be used similarily. They also include the ability to `POST` metadata changes to the data portal by allowing the validation and submission of change sheets or JSON files. Change sheets are spreadsheets that specify  ...
```
Same Q : "They also include the ability to `POST` metadata changes to the data portal by allowing the validation and submission of change sheets or JSON files" if this is user focused, should we leave this out? It could get confusing to users... make them think they can post / edit metadata... but they can't ... if we want to leave it in.. so it's 1 comprehensive read the docs for developers and users, maybe add throughout "As an NMDC developer, you can "

mslarae13 replied:
++ separate out the change sheets to a "for developers" section?

mslarae13 reviewed (commented):
I didn't really look for typos but have a few sentence re-work suggestions.

Also, we should consider who the documentation is for. I know we discussed it, but I don't recall the decision. And if the decision is both users & developers, we should separate out some of the developer / editing stuff to be more clear about who can have those permissions.. sections?

brynnz22 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+<br/>
+
+![find get activities](../_static/images/howto_guides/api_gui/find_get_activities.png)
+The `GET /activities` endpoint is a general way to fetch metadata about various activities (e.g. metagenome assembly, natural organic matter analysis, libarary preparation etc). Any "slot" (aka attribute) for [WorkflowExecutionActivty](https://microbiomedata.github.io/nmdc-schema/WorkflowExecutionActivity/) or [PlannedProcess](https://microbiomedata.github.io/nmdc-schema/PlannedProcess/)  may be  ...
```
Thanks @eecavanna for asking to clarify this. I updated this https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/e22bbfceddc2475177bfba820c9fb6575bf7fa9f. Let me know if it still needs some editing. I want it to make sense!

brynnz22 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+![find example step 1](../_static/images/howto_guides/api_gui/find_example_step1.png)
+2. Click **Try it out** in the upper right of the expanded endpoint box
+![find example step 2](../_static/images/howto_guides/api_gui/find_example_step2.png)
+3. Enter in parameters. In this case, we will input `funding_sources.search:EMSL` into the **filter** parameter. The `.search` performs a full text search to find studies with `funding_sources` that have the word "EMSL" in its value.
```
I added two sentences, one sentence at the end of each parameter table for specifying the optionality of the parameter if its missing the red `* required` . See commit b514ad4 Thanks!

brynnz22 commented on docs/howto_guides/api_gui.md:
```diff
@@ -0,0 +1,161 @@
+3. Enter in parameters. In this case, we will input `funding_sources.search:EMSL` into the **filter** parameter. The `.search` performs a full text search to find studies with `funding_sources` that have the word "EMSL" in its value.
+4. Click **Execute**
+![find example step 3 and step 4](../_static/images/howto_guides/api_gui/find_example_step3and4.png)
+5. View the results in a json format, available to download by clicking **Download**, or copy the results by clicking the clipboard icon in the bottom right corner of the response. In this case two studies were retrieved.
```
I also added a comma in a similar sentence. See commit [8388514](https://github.com/microbiomedata/NMDC_documentation/pull/27/commits/8388514bd0daf63e1a12cb8ffb03130d8e24a645)

brynnz22 commented:
@mslarae13 @dwinston @eecavanna @aclum I Incorporated all of your suggestions. Thank you all for your feedback! It was very helpful. I removed all of the `POST` endpoints regarding change sheets since after discussion with @mslarae13 this might be confusing for users who can't actually submit change sheets. The [documentation] that @dwinston (https://microbiomedata.github.io/nmdc-runtime/howto-guides/author-changesheets/) created for developers includes that information. I also implemented a spell checker so the typos should be fixed. Apologies for not doing this sooner, but thanks everyone for catching those! Please let me know of any other suggestions/feedback before I finalize this.

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -55,7 +58,7 @@ If the data object identifier is known, the metadata can be retrieved using the
 
 ![find get activities](../_static/images/howto_guides/api_gui/find_get_activities.png)
-The `GET /activities` endpoint is a general way to fetch metadata about various activities (e.g. metagenome assembly, natural organic matter analysis, library preparation, etc.). Any "slot" (a.k.a. attribute) for [WorkflowExecutionActivty](https://microbiomedata.github.io/nmdc-schema/WorkflowExecutionActivity/) or [PlannedProcess](https://microbiomedata.github.io/nmdc-schema/PlannedProcess/)  may ...
+The `GET /activities` endpoint is a general way to fetch metadata about various activities (e.g. metagenome assembly, natural organic matter analysis, library preparation, etc.). Any "slot" (a.k.a. attribute) for [WorkflowExecutionActivity](https://microbiomedata.github.io/nmdc-schema/WorkflowExecutionActivity/) or [PlannedProcess](https://microbiomedata.github.io/nmdc-schema/PlannedProcess/) cla ...
```
Thanks for updating this, @brynnz22! I do understand it now. 

I think I know where my confusion had been coming from: it says "slots for" and "attributes for" (as in "for"  a recipient/destination), whereas I think of things as "slots of" and "attributes of" (as in, "of" an owner). I think I would find it more clear with "of" instead of "for".

```diff
- Any "slot" (a.k.a. attribute) for [WorkflowExecutionActivity](https://microbiomedata.github.io/nmdc-schema/WorkflowExecutionActivity/)
+ Any "slot" (a.k.a. attribute) of [WorkflowExecutionActivity](https://microbiomedata.github.io/nmdc-schema/WorkflowExecutionActivity/)
```

```diff
- including attributes for subclasses of `WorkflowExecutionActivity`
+ including attributes of subclasses of `WorkflowExecutionActivity`
```

brynnz22 replied:
Thanks @eecavanna I fixed this!

eecavanna commented on docs/howto_guides/api_gui.md:
```diff
@@ -83,43 +86,21 @@ For more information and to see more examples of __find__ endpoints outside of t
+The [metadata endpoints](https://api.microbiomedata.org/docs#/metadata) can be used to get and filter metadata from collection set types (including studies, biosamples, activities, and data objects as discussed in the __find__ section). 
 
-The syntax for the filter parameter of the __metadata__ endpoints is slightly different than that of the __find__ endpoints using [MongoDB-like language querying](https://www.mongodb.com/docs/manual/tutorial/query-documents/) instead of the compact syntax the __find__ endpoints use. The applicable parameters of the __metadata__ endpoints, with acceptable syntax and examples, are in the table belo ...
+Unlike the compact syntax used in the __find__  endpoints, the syntax for the filter parameter of the metadata endpoints use [MongoDB-like language querying](https://www.mongodb.com/docs/manual/tutorial/query-documents/). The applicable parameters of the __metadata__ endpoints, with acceptable syntax and examples, are in the table below.
```
```diff
- the syntax ... use
+ the syntax ... uses
```

brynnz22 replied:
Thanks! Fixed!
